File: app/src/main/java/info/nightscout/androidaps/plugins/PumpDanaRS/comm/DanaRS_Packet_Bolus_Set_Step_Bolus_Start.java
package info.nightscout.androidaps.plugins.PumpDanaRS.comm;

import com.cozmo.danar.util.BleCommandUtil;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import info.nightscout.androidaps.Config;
import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.interfaces.Constraint;
import info.nightscout.androidaps.plugins.Overview.notifications.Notification;

public class DanaRS_Packet_Bolus_Set_Step_Bolus_Start extends DanaRS_Packet {
    private static Logger log = LoggerFactory.getLogger(DanaRS_Packet_Bolus_Set_Step_Bolus_Start.class);

    private double amount;
    private int speed;

    public boolean failed;
    public static int errorCode;

    public DanaRS_Packet_Bolus_Set_Step_Bolus_Start() {
        super();
        opCode = BleCommandUtil.DANAR_PACKET__OPCODE_BOLUS__SET_STEP_BOLUS_START;
    }

    // Speed 0 => 12 sec/U, 1 => 30 sec/U, 2 => 60 sec/U
    public DanaRS_Packet_Bolus_Set_Step_Bolus_Start(double amount, int speed) {
        this();

        // HARDCODED LIMIT
        amount = MainApp.getConstraintChecker().applyBolusConstraints(new Constraint<>(amount)).value();

        this.amount = amount;
        this.speed = speed;

        if (Config.logDanaMessageDetail)
            log.debug("Bolus start : " + amount + " speed: " + speed);
    }

    @Override
    public byte[] getRequestParams() {
        int stepBolusRate = (int) (amount * 100);
        byte[] request = new byte[3];
        request[0] = (byte) (stepBolusRate & 0xff);
        request[1] = (byte) ((stepBolusRate >>> 8) & 0xff);
        request[2] = (byte) (speed & 0xff);
        return request;
    }

    @Override
    public void handleMessage(byte[] data) {
        errorCode = intFromBuff(data, 0, 1);
        if (Config.logDanaMessageDetail) {
            if (errorCode == 0) {
                log.debug("Result OK");
                failed = false;
            } else {
                failed = true;
                log.error("Result Error: " + errorCode);
            }
        }
    }

    @Override
    public String getFriendlyName() {
        return "BOLUS__SET_STEP_BOLUS_START";
    }
}

File: danars/src/main/java/info/nightscout/androidaps/danars/comm/DanaRS_Packet_Bolus_Set_Step_Bolus_Start.kt
package info.nightscout.androidaps.danars.comm

import dagger.android.HasAndroidInjector
import info.nightscout.androidaps.interfaces.Constraint
import info.nightscout.androidaps.logging.LTag
import info.nightscout.androidaps.plugins.configBuilder.ConstraintChecker
import info.nightscout.androidaps.dana.DanaPump
import info.nightscout.androidaps.danars.encryption.BleEncryption
import javax.inject.Inject

class DanaRS_Packet_Bolus_Set_Step_Bolus_Start(
    injector: HasAndroidInjector,
    private var amount: Double = 0.0,
    private var speed: Int = 0
) : DanaRS_Packet(injector) {

    @Inject lateinit var danaPump: DanaPump
    @Inject lateinit var constraintChecker: ConstraintChecker

    init {
        opCode = BleEncryption.DANAR_PACKET__OPCODE_BOLUS__SET_STEP_BOLUS_START
        // Speed 0 => 12 sec/U, 1 => 30 sec/U, 2 => 60 sec/U
        // HARDCODED LIMIT - if there is one that could be created
        amount = constraintChecker.applyBolusConstraints(Constraint(amount)).value()
        aapsLogger.debug(LTag.PUMPCOMM, "Bolus start : $amount speed: $speed")
    }

    override fun getRequestParams(): ByteArray {
        val stepBolusRate = (amount * 100).toInt()
        val request = ByteArray(3)
        request[0] = (stepBolusRate and 0xff).toByte()
        request[1] = (stepBolusRate ushr 8 and 0xff).toByte()
        request[2] = (speed and 0xff).toByte()
        return request
    }

    override fun handleMessage(data: ByteArray) {
        danaPump.bolusStartErrorCode = intFromBuff(data, 0, 1)
        if (danaPump.bolusStartErrorCode == 0) {
            failed = false
            aapsLogger.debug(LTag.PUMPCOMM, "Result OK")
        } else {
            aapsLogger.error("Result Error: ${danaPump.bolusStartErrorCode}")
            failed = true
        }
    }

    override fun getFriendlyName(): String {
        return "BOLUS__SET_STEP_BOLUS_START"
    }
}